knimin/lib/data_access.py

```diff
@@ -3165,96 +3031,130 @@ def read_sample_plate(self, plate_id):
         with TRN:
-            self._sample_plate_exists(id)
-            if self._sample_plate_layout_exists(id):
-                self._clear_sample_plate_layout(id)
-            sql = """INSERT INTO pm.sample_plate_layout (sample_plate_id,
-                        sample_id, col, row, name, notes)
+            self._sample_plate_exists(sample_plate_id)
+
+            # Check if the given layout forms a valid layout
+            l_rows = len(layout)
+            l_cols = len(layout[0])
+            # Check that all rows have the same number of columns
+            # for r in layout
+            if any([len(r) != l_cols for r in layout]):
+                raise ValueError(
+                    "The given layout doesn't form a valid plate map because "
+                    "not all rows have the same number of columns")
+
+            # Get the sample plate type size
+            sql = """SELECT rows, cols
+                     FROM pm.plate_type
+                        JOIN pm.sample_plate USING (plate_type_id)
+                    WHERE sample_plate_id = %s"""
+            TRN.add(sql, [sample_plate_id])
+            # Magic number 0 -> there is only one result on the previous query
+            plate_dims = dict(TRN.execute_fetchindex()[0])
+
+            # Check that the given layout has the same dimensions as the
+            # plate type
+            if plate_dims['rows'] != l_rows or plate_dims['cols'] != l_cols:
+                raise ValueError(
+                    "The given layout doesn't match the plate type "
+                    "dimensions. Plate type: (%d, %d). Layout: (%d, %d)"
+                    % (plate_dims['rows'], plate_dims['cols'], l_rows, l_cols))
+
+            # We can set the new layout. First clear the previous layout
+            # and insert the new one
+            self._clear_sample_plate_layout(sample_plate_id)
+            sql = """INSERT INTO pm.sample_plate_layout
+                        (sample_plate_id, sample_id, row, col, name, notes)
                      VALUES (%s, %s, %s, %s, %s, %s)"""
-            for x in layout:
-                sample_id = x['sample_id']
-                if not self._sample_exists(sample_id):
-                    raise ValueError('Sample ID %s does not exist.'
-                                     % sample_id)
-                TRN.add(sql, [id, sample_id, x['col'], x['row'],
-                              x.get('name'), x.get('notes')])
-                TRN.execute()
+            sql_args = []
+            for row_idx, row in enumerate(layout):
+                for col_idx, values in enumerate(row):
+                    sql_args.append([sample_plate_id, values['sample_id'],
+                                     row_idx, col_idx, values['name'],
+                                     values['notes']])
+            TRN.add(sql, sql_args, many=True)
+            TRN.execute()
 
-    def read_sample_plate_layout(self, id):
+    def read_sample_plate_layout(self, sample_plate_id):
```

Comment:
Note that the representation changed. Instead of being a list of dicts with 'col', 'row', 'sample', etc, the current representation is just a dense 2D matrix storing this values. I think this representation makes more sense given than in general the plates tend to be dense rather than sparse, so a matrix representation is more efficient.